mesonbuild/dependencies/base.py

```diff
@@ -640,3 +640,45 @@ def find_external_dependency(name, env, kwargs):
         raise pkg_exc
     mlog.log('Dependency', mlog.bold(name), 'found:', mlog.red('NO'))
     return pkgdep
+
+
+class _StripSystemLibdirs:
+
+    """Remove -L<system path> arguments.
+
+    leaving these in will break builds where a user has a version of a library
+    in the system path, and a different version not in the system path if they
+    want to link against the non-system path version.
+    """
+
+    def __init__(self):
+        # FIXME: For BSD systems we may need /usr/local/lib* in here too, but
+        # we don't want those values for linux where /usr/local/lib is not
+        # normally a system library path
+        system_paths = ['/lib', '/lib32', '/lib64', '/usr/lib', '/usr/lib32',
+                        '/usr/lib64']
+
+        # Debian introduces a whole nother level of complexity, they also have
+        # /lib/<debian-tuple> and /usr/lib/<debian-tuple>. There's a list of
+        # them here: https://wiki.debian.org/Multiarch/Tuples
+        #
+        # Rather than trying to list all of them here, we can use dpkg to check
+        # what the tripplet should be
+        if os.path.exists('/etc/debian_version'):
+            p, out = Popen_safe(['dpkg-architecture', '-q', 'DEB_HOST_MULTIARCH'])[:2]
+            if p.returncode != 0:
+                mlog.log(mlog.yellow('WARNING:'),
+                         'could not run dpkg-architecture, system -L paths may'
+                         ' be present in linker. This may cause undesired behavior')
+            else:
+                path = out.strip()
+                system_paths += os.path.join('/lib', path)
+                system_paths += os.path.join('/usr/lib', path)
```

Comment:
This works for native building but what about cross compilation where the paths are different?

Reply:
You're right, it wouldn't work for cross compiling. Would it be better to just harcode all of the debian paths in, or would it be better to try to convert meson's target machine values into something that dpkg-architecture can gives us a path for?

Reply:
It looks like both `gcc` and `clang` support `-print-search-dirs` to print out their idea of the system library path, which also takes any `sysroot` args into account. On the other hand, since this is per-compiler, there'd need to be an additional filter stage combining compiler + dependency.

Reply:
It looks like fooishbar's suggestion will work for *nix and osx. Is there the concept of a "system path" on windows?